Dev/adi axi tdd

eliza-balas opened the pull request.

eliza-balas commented:
I will make the required changes after the CI checks are done.

eliza-balas commented:
I will resolve the dt_binding_check error check, but until then, a review for the rest of the code would be useful.

amiclaus commented on MAINTAINERS:
```diff
@@ -1351,6 +1351,14 @@ F:	drivers/iio/amplifiers/hmc425a.c
 X:	drivers/iio/*/adjd*
 
+ANALOG DEVICES INC GENERIC TDD ENGINE DRIVER
+M:	David Winter <david.winter@analog.com>
```
this should go with the dt-bindings

amiclaus commented on drivers/misc/adi-axi-tdd.c:
```diff
@@ -0,0 +1,862 @@
+
+#define ADI_REG_TDD_INTERFACE_DESCRIPTION 0x0010
+#define ADI_TDD_SYNC_COUNT_WIDTH GENMASK(30, 24)
+#define ADI_TDD_SYNC_COUNT_WIDTH_GET(x) FIELD_GET(ADI_TDD_SYNC_COUNT_WIDTH, x)
```
use FIELD_GET inline and remove all these macros.

amiclaus commented on drivers/misc/adi-axi-tdd.c:
```diff
@@ -0,0 +1,862 @@
+
+/* Register Map */
+
+#define ADI_REG_TDD_MAGIC 0x000c
```
please properly indent all the macro definitions

amiclaus commented on drivers/misc/adi-axi-tdd.c:
```diff
@@ -0,0 +1,862 @@
+				 ADI_AXI_PCORE_VER_MINOR(data),
+				 ADI_AXI_PCORE_VER_PATCH(data));
+		break;
+
```
remove all these empty lines after break;

amiclaus commented on drivers/misc/adi-axi-tdd.c:
```diff
@@ -0,0 +1,862 @@
+static ssize_t adi_axi_tdd_show(struct device *dev,
+				struct device_attribute *dev_attr, char *buf)
+{
+	struct adi_axi_tdd_state *st = dev_get_drvdata(dev);
```
please properly arrange all these variable declarations/definitions in reverse x-mas order. check the entire driver for similar issues.

amiclaus commented on drivers/misc/adi-axi-tdd.c:
```diff
@@ -0,0 +1,862 @@
+		ret = count;
+		tdd_write(st, ADI_AXI_REG_SCRATCH, data);
+		break;
+
```
ditto.

amiclaus commented on drivers/misc/adi-axi-tdd.c:
```diff
@@ -0,0 +1,862 @@
+	case ADI_TDD_ATTR_SCRATCH:
+		ret = kstrtou32(buf, 0, &data);
+		if (ret)
+			break;
```
we can return directly here. same for all other cases.

amiclaus commented on drivers/misc/adi-axi-tdd.c:
```diff
@@ -0,0 +1,862 @@
+	ret = adi_axi_tdd_init_synthesis_parameters(st);
+	if (ret) {
+		dev_err(&pdev->dev,
+			"Failed to load synthesis parameters, make sure the device is configured correctly.\n");
```
return ret.

amiclaus commented on drivers/misc/adi-axi-tdd.c:
```diff
@@ -0,0 +1,862 @@
+	case ADI_TDD_ATTR_CHANNEL_ON_RAW:
+		data = tdd_read(st, ADI_REG_TDD_CHANNEL(channel, ADI_TDD_CHANNEL_ON));
+		if (ms)
+			ret = adi_axi_tdd_format_ms(st, data, buf);
```
return directly?

amiclaus commented on drivers/misc/adi-axi-tdd.c:
```diff
@@ -0,0 +1,862 @@
+		if (ms)
+			ret = adi_axi_tdd_format_ms(st, data, buf);
+		else
+			ret = sysfs_emit(buf, "%u\n", data);
```
return directly? actually this applies to most of the cases.

amiclaus commented on Documentation/devicetree/bindings/misc/adi,axi-tdd.yaml:
```diff
@@ -0,0 +1,43 @@
+            clocks = <&zynqmp_clk PL0_REF>, <&zynqmp_clk PL1_REF>;
+            clock-names = "s_axi_aclk", "intf_clk";
+    };
+
```
nit: extra line

amiclaus commented on arch/arm64/boot/dts/xilinx/zynqmp-zcu102-rev10-tdd.dts:
```diff
@@ -0,0 +1,41 @@
+			clocks = <&zynqmp_clk PL1_REF>, <&zynqmp_clk PL0_REF>;
+			clock-names = "intf_clk", "s_axi_aclk";
+		};
+
```
nit: extra line

amiclaus commented on arch/arm64/boot/dts/xilinx/zynqmp-zcu102-rev10-tdd.dts:
```diff
@@ -0,0 +1,41 @@
+/*
+ * Analog Devices TDD Test Device Tree
+ *
+ * Copyright (C) 2022 Analog Devices Inc.
```
2023?

amiclaus commented on MAINTAINERS:
```diff
@@ -1351,6 +1351,14 @@ F:	drivers/iio/amplifiers/hmc425a.c
+M:	Eliza Balas <eliza.balas@analog.com>
+S:	Supported
+W:	http://wiki.analog.com/resources/fpga/docs/axi_tdd
+W:	http://ez.analog.com/community/linux-device-drivers
```
please use the correct link: https://ez.analog.com/linux-software-drivers/

amiclaus commented on Documentation/devicetree/bindings/misc/adi,axi-tdd.yaml:
```diff
@@ -0,0 +1,43 @@
+  - clocks
+  - clock-names
+
+additionalProperties: false
```
i think `unevaluatedProperties` is preferred over `additionalProperties`

amiclaus commented on Documentation/devicetree/bindings/misc/adi,axi-tdd.yaml:
```diff
@@ -0,0 +1,43 @@
+
+description: |
+  Bindings for the new generic TDD CORE, which is part of the Analog Devices hdl reference designs.
+  For more information see the wiki: https://wiki.analog.com/resources/fpga/docs/axi_tdd
```
add new line.

amiclaus commented on Documentation/devicetree/bindings/misc/adi,axi-tdd.yaml:
```diff
@@ -0,0 +1,43 @@
+# SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause)
+# Copyright 2021 Analog Devices Inc.
```
2023

amiclaus commented on drivers/misc/adi-axi-tdd.c:
```diff
@@ -0,0 +1,862 @@
+/*
+ * TDD HDL CORE driver
+ *
+ * Copyright 2022 Analog Devices Inc.
```
2023

amiclaus reviewed (commented):
some initial comments on my side. Also the commit order should be:
1. dt-bindings
2. driver support
3. Kconfig.adi
4. dts

eliza-balas commented:
Version 2

* resolved dt_binding_check error
* added clocks description and clock-names const in dt_binding
* added unevaluatedProperties instead of additionalProperties in dt_binding
* changed year to 2023 in Copyright
* replaced link to http://ez.analog.com/linux-software-drivers/ in MAINTAINERS file
* added MAINTAINERS file to dt_binding commit
* indented all the macro definitions in driver
* added return for all read/write functions in driver
* added if functions to check for errors for returns in case functions in driver
* removed new lines after breaks in case functions in driver
* replaced 1ULL shifts with BIT_ULL in driver
* replaced 0xFFFFFFFFU with GENMASK_ULL() in driver
* added FIELD_GET / FIELD_PREP where needed instead of macros in driver

amiclaus commented on arch/arm64/boot/dts/xilinx/zynqmp-zcu102-rev10-tdd.dts:
```diff
@@ -0,0 +1,41 @@
+ * Analog Devices TDD Test Device Tree
+ *
+ * Copyright (C) 2022 Analog Devices Inc.
+ */
```
The dt_build error seems to be valid. Please add something similar to:
https://github.com/analogdevicesinc/linux/blob/7c2858b8b6009c0be4fbda434330e965e6464f03/arch/arm/boot/dts/zynq-zed-adv7511-adf4377.dts#L6

eliza-balas commented:
Version 3

* changed year to 2023 for Copyright in dts example file
* added hdl_project name in dts example file

nunojsa commented on drivers/misc/adi-axi-tdd.c:
```diff
@@ -0,0 +1,845 @@
+
+static struct platform_driver adi_axi_tdd_driver = {
+	.driver = { .name = "adi_axi_tdd",
+		    .of_match_table = adi_axi_tdd_of_match },
```
nit: please look at  other driver to see how this looks like in terms of style...

also, `of_match_table` should be right above...

nunojsa commented on drivers/misc/adi-axi-tdd.c:
```diff
@@ -0,0 +1,845 @@
+
+	ret = adi_axi_tdd_init_sysfs(pdev, st);
+	if (ret) {
+		dev_err(&pdev->dev, "Failed to init sysfs, aborting ...\n");
```
use `dev_err_probe()`... ditto for all other places in the probe path

nunojsa commented on drivers/misc/adi-axi-tdd.c:
```diff
@@ -0,0 +1,845 @@
+	if (IS_ERR(st->regs))
+		return PTR_ERR(st->regs);
+
+	ret = tdd_read(st, ADI_AXI_REG_VERSION, &version);
```
hmm, you are calling `devm_regmap_init_mmio()` and then not using regmap API :)?

eliza-balas replied:
the tdd_read function returns regmap_read function

nunojsa commented on drivers/misc/adi-axi-tdd.c:
```diff
@@ -0,0 +1,845 @@
+		}
+	}
+
+/* Terminate list, technically not necessary because kzalloc was used */
```
indent this comment

nunojsa commented on drivers/misc/adi-axi-tdd.c:
```diff
@@ -0,0 +1,845 @@
+		ret = tdd_write(st, ADI_REG_TDD_CHANNEL(channel, ADI_TDD_CHANNEL_OFF), (u32)data64);
+		break;
+	default:
+		// Must not happen
```
remove the comment

pcercuei replied:
If it is actually impossible for it to happen, you can use `unreachable();`

nunojsa commented on drivers/misc/adi-axi-tdd.c:
```diff
@@ -0,0 +1,845 @@
+			return adi_axi_tdd_format_ms(st, data, buf);
+		return sysfs_emit(buf, "%u\n", data);
+	default:
+		// Must not happen
```
ditto

nunojsa commented on drivers/misc/adi-axi-tdd.c:
```diff
@@ -0,0 +1,845 @@
+	if (len >= PAGE_SIZE - 1)
+		return -EFBIG;
+
+	return len + sprintf(buf + len, "\n");
```
can't this be done with `sysf_emit()` and `sysfs_emit_at()`?

nunojsa commented on drivers/misc/adi-axi-tdd.c:
```diff
@@ -0,0 +1,845 @@
+	}
+
+	*integer = i;
+	*fract = f;
```
hmm this is a copy from what is done in IIO... maybe you can purpose to upstream to have a generic API in lib/?

eliza-balas replied:
I will try to talk with someone who maintains lib/ about this

nunojsa commented on drivers/misc/adi-axi-tdd.c:
```diff
@@ -0,0 +1,845 @@
+			   const u32 mask, const u32 val)
+{
+	return regmap_update_bits_base(st->regs, reg, mask, val, NULL, false, false);
+}
```
hmmm I see... However, I don't really the point for these function. Just use plain regmap_update_bits(). Ditto for tdd_read() and tdd_write().

nunojsa commented on drivers/misc/adi-axi-tdd.c:
```diff
@@ -0,0 +1,845 @@
+	ret = regmap_read(st->regs, lreg, &data);
+	if (ret)
+		return ret;
+	*val |= FIELD_PREP(GENMASK_ULL(31, 0), data);
```
the 64 variants need a mutex to protect against concurrent calls... Note you have two reads and two writes

eliza-balas commented on drivers/misc/adi-axi-tdd.c:
```diff
@@ -0,0 +1,845 @@
+	struct adi_axi_tdd_clk *clk = data;
+
+	clk_notifier_unregister(clk->clk, &clk->nb);
+}
```
sorry for my misunderstanding, these are callbacks used in devm_add_action_or_reset function. Do you mean to move them closer to probe() function for better readability?

nunojsa replied:
> sorry for my misunderstanding, these are callbacks used in devm_add_action_or_reset function. Do you mean to move them closer to probe() function for better readability?

yes, and also reducing their scope. It's a nitpick, I know, but it is a good practice...

eliza-balas commented:
Version 4

* removed IIO dependency and added `select REGMAP_MMIO` in Kconfig
* removed unused math64.h and string.h header files in driver
* removed tdd_write, tdd_read  and tdd_update_bits functions in driver
* replaced scnprintf and sprintf with sysfs_emit and sysfs_emit_at in driver
* used unreachable() for default cases in switch statements in driver
* removed adi_axi_tdd_clk_setup function and added its functionality in probe function in driver

eliza-balas commented:
Version 5

* removed adi_axi_tdd_str_to_fixpoint function in driver
* modified adi_axi_tdd_parse_ms function in driver to return success or error

Demon000 commented on drivers/misc/adi-axi-tdd.c:
```diff
@@ -0,0 +1,817 @@
+	mutex_lock(&st->lock);
+	ret = regmap_read(st->regs, hreg, &data);
+	if (ret)
+		return ret;
```
Missing mutex unlock here. Add an `exit` label before the unlock, and rather than returning here, do `goto exit;`. Same for below.

Demon000 replied:
Missing mutex unlock here. Add an `exit` label before the unlock, and rather than returning here, do `goto exit;`. Same for below.

Demon000 commented on drivers/misc/adi-axi-tdd.c:
```diff
@@ -0,0 +1,817 @@
+	mutex_lock(&st->lock);
+	ret = regmap_write(st->regs, lreg, FIELD_GET(GENMASK_ULL(31, 0), val));
+	if (ret)
+		return ret;
```
Same as below.

Demon000 commented on drivers/misc/adi-axi-tdd.c:
```diff
@@ -0,0 +1,817 @@
+			return adi_axi_tdd_format_ms(st, data, buf);
+		return sysfs_emit(buf, "%u\n", data);
+	default:
+		unreachable();
```
Just return `-EINVAL`.

Demon000 commented on drivers/misc/adi-axi-tdd.c:
```diff
@@ -0,0 +1,817 @@
+		ret = kstrtou32(buf, 0, &data);
+		if (ret)
+			return ret;
+		ret = regmap_write(st->regs, ADI_AXI_REG_SCRATCH, data);
```
We usually try to return out of a switch case where possible, especially when the switch has a lot of branches.

What you could do is move the entire contents of this function to a new function, where you do `return regmap_write()` / `return regmap_update_bits_base()` in each switch case. Then, from this store function, call that function, and then do `return ret ?: count` at the end.

Demon000 commented on drivers/misc/adi-axi-tdd.c:
```diff
@@ -0,0 +1,817 @@
+	if (!st->burst_count_width || !st->reg_width)
+		return -EINVAL;
+
+	st->sync_count_mask = BIT_ULL(st->sync_count_width) - 1ULL;
```
GENMASK?

Demon000 commented on drivers/misc/adi-axi-tdd.c:
```diff
@@ -0,0 +1,817 @@
+/* Match table for of_platform binding */
+static const struct of_device_id adi_axi_tdd_of_match[] = {
+	{ .compatible = "adi,axi-tdd-2.00.a" },
+	{}
```
Nitpick: space between `{ }`.

nunojsa commented on drivers/misc/adi-axi-tdd.c:
```diff
@@ -0,0 +1,803 @@
+	u32 i, j;
+
+	channel_attributes = devm_kzalloc(&pdev->dev,
+					  6 * st->channel_count * sizeof(*channel_attributes), GFP_KERNEL);
```
this line looks to be big... isn't it more than the 100 limit?

nunojsa commented on drivers/misc/adi-axi-tdd.c:
```diff
@@ -0,0 +1,803 @@
+	tdd_attrs = devm_kzalloc(&pdev->dev,
+				 attribute_count * sizeof(*tdd_attrs), GFP_KERNEL);
+	if (!tdd_attrs)
+		return -ENOMEM;
```
Consider using `devm_kcalloc()` for the above two calls

nunojsa commented on drivers/misc/adi-axi-tdd.c:
```diff
@@ -0,0 +1,803 @@
+	}
+
+	/* Terminate list, technically not necessary because kzalloc was used */
+	tdd_attrs[base_attr_count + 6 * st->channel_count] = NULL;
```
then, maybe not do this?

nunojsa commented on drivers/misc/adi-axi-tdd.c:
```diff
@@ -0,0 +1,803 @@
+};
+module_platform_driver(adi_axi_tdd_driver);
+
+MODULE_AUTHOR("David Winter <david.winter@analog.com>");
```
Maybe put yourself also as module author... You can have multiple `MODULE_AUTHOR()` lines

nunojsa reviewed (changes requested):
"´The exposed sysfs interface follows the IIO specification, but is not registered into the IIO device tree, and does not rely on IIO kernel APIs. This means the device will now show up in IIO by default, but may be made available using an iio_fakedev"

I would remove these lines from the Kconfig entry and the commit messages... I mean, you can say the ABI resembles a bit the IIO one (when dealing with channels) and even bring the question if this could ever be considered an IIO  device when upstreaming this (I suspect it can't but who knows...). However, the part where you mention `iio_fakedev` is pretty much ADI internal stuff so I would not put it in here (at least on the Kconfig). You can leave it in the commit messages if you really want as long as you remember to change the messages when upstreaming this driver,

mhennerich commented:
BTW - do we have started with a wiki page for this new driver yet?

eliza-balas commented:
> BTW - do we have started with a wiki page for this new driver yet?

We have a wiki page for the TDD core: https://wiki.analog.com/resources/fpga/docs/axi_tdd
The wiki page will include a section about the Linux driver, similar to what was specified in the previous version of the wiki page ([https://wiki.analog.com/resources/fpga/docs/axi_tdd?rev=1634050498](https://wiki.analog.com/resources/fpga/docs/axi_tdd?rev=1634050498))
The section will be added as soon as we have a final version of the driver and is accepted upstream.

nunojsa commented:
Hi @eliza-balas,

I'm aware that we won't be able to upstream this driver now and will have to merge it as-is. Can you rebase onto master? This is still 5.15 based and master is now 6.1. I think you might not need your last patch anymore.

eliza-balas commented:
> Hi @eliza-balas,
> 
> I'm aware that we won't be able to upstream this driver now and will have to merge it as-is. Can you rebase onto master? This is still 5.15 based and master is now 6.1. I think you might not need your last patch anymore.

Yes, I can do that, if @mhennerich is okay with this.

nunojsa commented on drivers/misc/adi-axi-tdd.c:
```diff
@@ -692,14 +697,34 @@ static int adi_axi_tdd_probe(struct platform_device *pdev)
+				       st->clk.clk);
+	if (ret)
+		return ret;
+
```
you should not need this patch anymore. [devm_clk_get_enabled()](https://elixir.bootlin.com/linux/v6.1/C/ident/devm_clk_get_enabled) is already in 6.1

mhennerich commented:
> Yes, I can do that, if @mhennerich is okay with this.

Ok - with me. Nuno will backport this to 5.15 next_stable

eliza-balas commented:
> Ok - with me. Nuno will backport this to 5.15 next_stable

@nunojsa let me know when you will add this to next_stable, so I can make the rebase for 6.1 and remove the COMPAT commit.

nunojsa commented:
> @nunojsa let me know when you will add this to next_stable, so I can make the rebase for 6.1 and remove the COMPAT commit.

I'm actually waiting on you to first have this merged in master :sweat_smile:

eliza-balas reviewed (commented):
merging this to master
